### PR TITLE
Android Embedding Refactor PR31: Integrate platform views with the new embedding and the plugin shim.

### PR DESCRIPTION
**Pull request by matthew-carroll**
Android Embedding Refactor PR31: Integrate platform views with the new embedding and the plugin shim.

To test this change, I ported the WebView plugin locally to use the new Android embedding. It works as expected.

This PR adjusts where and when the `PlatformViewsController` is created and attached to an `Activity` to respect the possibility that an `Activity` does not initially exist. It does so in a way that still allows the `ShimPluginRegistry` to work as expected with existing plugins in a full-Flutter scenario.

Additional investigation and work on `PlatformViewsController` is still necessary, but that future work is not expected to require fundamental changes to the embedding APIs, so I've left it for future PRs. With this PR, we should be able to say that platform views are integrated with the new embedding, but without leaking an `Activity` and without crashing some bots.

This PR should constitute the completion of known requirements for the core embedding + plugin API (minus any future decisions about plugin registration, etc.).

### CODE REVIEW

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java [hunk 1]
```diff
@@ -112,6 +116,8 @@ public void onFirstFrameRendered() {
    *   <li>{@link #renderMode} defaults to {@link RenderMode#surface}.</li>
    *   <li>{@link #transparencyMode} defaults to {@link TransparencyMode#opaque}.</li>
    * </ul>
+   * {@code FlutterView} requires an {@code Activity} instead of a generic {@code Context}
```
Comment (mklim) [inline shell/platform/android/io/flutter/embedding/android/FlutterView.java]: Should the signature be changed to actually take an Activity throughout?

**Comment by xster** [comment]
Can you reference a bug for remaining work in the platform view theme?

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java [hunk 2]
```diff
@@ -66,7 +67,10 @@
  * {@code Fragment}.
  */
 // TODO(mattcarroll): explain each call forwarded to Fragment (first requires resolution of PluginRegistry API).
-public class FlutterActivity extends FragmentActivity implements OnFirstFrameRenderedListener {
+public class FlutterActivity extends FragmentActivity
+    implements FlutterFragment.FlutterEngineProvider,
+    FlutterFragment.FlutterEngineConfigurator,
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: I'm reading the PR sequentially so I'm making comments based on context from what's above the comment.

I'm getting the impression we were going to just push in and take away an activity from the platformviewcontroller. In that case, why is the activity involved? Couldn't the fragment just always get the activity on its own lifecycle events?

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java [hunk 3]
```diff
@@ -353,6 +357,25 @@ protected boolean shouldAttachEngineToActivity() {
     return true;
   }
 
+  /**
+   * Hook for subclasses to easily provide a custom {@code FlutterEngine}.
+   */
+  @Nullable
+  @Override
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: Not clear what's the relationship with these and getFlutterEngine.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 4]
```diff
@@ -296,6 +297,8 @@ protected Bundle createArgs() {
   private FlutterView flutterView;
   @Nullable
   private PlatformPlugin platformPlugin;
+  @NonNull
+  private final PlatformViewsController platformViewsController = new PlatformViewsController();
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: Are we moving the ownership of the platformviewscontroller out of the plugins registry into here?

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 5]
```diff
@@ -358,8 +361,14 @@ public void onAttach(Context context) {
       // which means there shouldn't be any possibility for the Fragment Lifecycle to get out of
       // sync with the Activity. We use the Fragment's Lifecycle because it is possible that the
       // attached Activity is not a LifecycleOwner.
-      flutterEngine.getActivityControlSurface().attachToActivity(getActivity(), getLifecycle());
+      flutterEngine.getActivityControlSurface().attachToActivity(
+          getActivity(),
+          getLifecycle(),
+          platformViewsController
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: if you own it, you should operate it too
e.g.

platformViewController.here'sTheActivity(getActivity())

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 6]
```diff
@@ -398,11 +407,11 @@ private void setupFlutterEngine() {
       // Defer to the Activity that owns us to provide a FlutterEngine.
       Log.d(TAG, "Deferring to attached Activity to provide a FlutterEngine.");
       FlutterEngineProvider flutterEngineProvider = (FlutterEngineProvider) attachedActivity;
-      flutterEngine = flutterEngineProvider.getFlutterEngine(getContext());
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: is this just an API swap?

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 7]
```diff
@@ -431,10 +440,33 @@ protected FlutterEngine createFlutterEngine(@NonNull Context context) {
...
+   * an {@link io.flutter.embedding.engine.plugins.activity.ActivityControlSurface}. In that
+   * case, any configuration, e.g., plugins, must not expect or depend upon an available
+   * {@code Activity} at the time that this method is invoked.
+   * <p>
+   * The default behavior of this method is to defer to the owning {@code FragmentActivity}
+   * as a {@link FlutterEngineConfigurator}. Subclasses can override this method if the
+   * subclass needs to override the {@code FragmentActivity}'s behavior, or add to it.
+   */
+  protected void configureFlutterEngine(@NonNull FlutterEngine flutterEngine) {
+    FragmentActivity attachedActivity = getActivity();
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: So this is an auxiliary API to give more control to the activity in case the fragment wants to give the activity to a platfromviewcontroller and the activity says no?

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 8]
```diff
@@ -538,7 +570,7 @@ public void onStart() {
     new Handler().post(new Runnable() {
       @Override
       public void run() {
-        flutterView.attachToFlutterEngine(flutterEngine);
+        flutterView.attachToFlutterEngine(flutterEngine, platformViewsController);
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: I'm surprised something needed to be done to the platform view controller multiple times

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 9]
```diff
@@ -783,6 +815,27 @@ protected void onFirstFrameRendered() {}
      * to provide its own {@code FlutterEngine} instance.
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: comment for line 811

> `Returns the {@link FlutterEngine} that should be used by a child {@code FlutterFragment}.` 

This is just an interface right? We don't know that it has a child per se.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java [hunk 10]
```diff
@@ -783,6 +815,27 @@ protected void onFirstFrameRendered() {}
...
      */
     @Nullable
-    FlutterEngine getFlutterEngine(@NonNull Context context);
+    FlutterEngine provideFlutterEngine(@NonNull Context context);
+  }
+
+  /**
+   * Configures a {@link FlutterEngine} after it is created, e.g., adds plugins.
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: Based on this description, it makes me assume that the default FlutterActivity would do something here but it just no-ops.

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java [hunk 11]
```diff
@@ -69,6 +71,8 @@
   @Nullable
   private FlutterRenderer.RenderSurface renderSurface;
   private boolean didRenderFirstFrame;
+  @Nullable
+  private PlatformViewsController platformViewsController;
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterView.java]: I understand why one is given to the flutterview on attachToFlutterEngine but I'm not sure I understand why the flutterview instance then holds onto it.

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java (hunk 1 above)
Comment (xster) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java]: I thought the approach we were talking about last week was that we're keeping the PlatformViewController to have an ApplicationContext constructor but just that we're adding new APIs to it, something like onAttachToActivity. And you can only do presentation.show after you onAttachToActivity. But then FlutterView isn't involved in any of this. It's just between the fragment, which does flutterPluginRegistry.getPlatformViewsController.onAttachActivity(activity), and the PlatformViewsController.

File: shell/platform/android/io/flutter/embedding/engine/FlutterEnginePluginRegistry.java [hunk 12]
```diff
@@ -259,16 +260,32 @@ private boolean isAttachedToActivity() {
...
+        activity,
+        platformViewsController
+    );
     this.flutterEngineAndroidLifecycle.setBackingLifecycle(lifecycle);
 
+    // Activate the PlatformViewsController. This must happen before any plugins attempt
+    // to use it, otherwise an error strack trace will appear that says there is no
+    // flutter/platform_views channel.
+    platformViewsController.attach(
+        activity,
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/engine/FlutterEnginePluginRegistry.java]: So this part I understand. I'm not sure why anyone besides the fragment is involved in FlutterEngineActivityPluginRegistry.attachToActivity.

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java (hunk 1 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java]: @mklim I tried that but compilation failed because apparently Android `View` subclasses need to declare exact replicas of superclass constructors (probably due to XML inflation concerns). So I backed out that change and left a comment.

@xster the reason this was a problem at all is because we can't use an application context. If you ever attempt to `show()` a dialog/presentation with an application context, the app will blow up. That's why the previous commit that used the application context had to be reverted, which sparked the conversation with the Android team, etc. So now we're doing what we have to do to give `PlatformViewsController` an `Activity`.

**Comment by matthew-carroll** [comment]
> Can you reference a bug for remaining work in the platform view theme?

I'm not sure what this is referring to?

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java (hunk 2 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: I'm not sure what you're asking here. Can you rephrase?

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java (hunk 4 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: Yes. As mentioned in the doc for the new plugin registry, its job is now exclusively to register plugins, not to initialize and hold arbitrary resources and dependencies.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java (hunk 5 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: I don't think that makes much sense. This is typical dependency injection. Notice that `FlutterFragment` doesn't "operate" its `Activity` or `Lifecycle`, it passes those as references to the things that actually work with them. `PlatformViewsController` is no different.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java (hunk 6 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: Yeah. I realized in this PR that there were 2 very similar methods called `getFlutterEngine` so I wanted to differentiate.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java (hunk 8 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: I'm not sure what you're referring to?

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java (hunk 9 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: It's an interface that is defined within `FlutterFragment` for use by `Activity`s that contain `FlutterFragment`s.

File: shell/platform/android/io/flutter/embedding/engine/FlutterEnginePluginRegistry.java (hunk 12 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/engine/FlutterEnginePluginRegistry.java]: Can you specify exactly what you mean? The reference to "anyone besides the fragment" leaves this ambiguous as to what you're getting at.

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java (hunk 11 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java]: How would we call `detachAccessibilityBridge()` without a reference to the `PlatformViewsController`?

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java (hunk 3 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: `getFlutterEngine` has been renamed to `provideFlutterEngine` and its purpose remains unchanged.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java (hunk 2 above)
Comment (xster) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: You were saying that this whole file's diff is unrelated to the platform view thing so disregard my comment.

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java [hunk 13]
```diff
@@ -511,7 +528,10 @@ private void resetWillNotDraw(boolean isAccessibilityEnabled, boolean isTouchExp
    * See {@link #detachFromFlutterEngine()} for information on how to detach from a
    * {@link FlutterEngine}.
    */
-  public void attachToFlutterEngine(@NonNull FlutterEngine flutterEngine) {
+  public void attachToFlutterEngine(
+      @NonNull FlutterEngine flutterEngine,
+      @NonNull PlatformViewsController platformViewsController
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterView.java]: So now if I am just integrating via a FlutterView, I have to create a PlatformViewController too?

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java [hunk 14]
```diff
@@ -537,7 +558,7 @@ public void attachToFlutterEngine(@NonNull FlutterEngine flutterEngine) {
     textInputPlugin = new TextInputPlugin(
         this,
         this.flutterEngine.getDartExecutor(),
-        null
+        platformViewsController
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterView.java]: I see. Can we flip it around? Such that the FlutterView creates and owns the PlatformViewController and since the FlutterFragment only needs this in one place, it does 

```java
flutterEngine.getActivityControlSurface().attachToActivity(
          getActivity(),
          getLifecycle(),
          flutterView.getPlatformViewsController(),
)
```

Regarding the sequencing of onAttach->onCreate->onCreateView etc, I think we might want to tweak it a bit too. Otherwise it's a bit leaky that we're given an Activity which hasn't yet finished initializing to an activity aware plugin. It's a bit like giving a reference to a partially constructed object to someone else in Android equivalent.

If the flutterEngine.getActivityControlSurface().attachToActivity() was on onActivityCreated() instead, then it's both cleaner and onCreateView would have been called by then, meaning you can get the PlatformViewController from the FlutterView.

File: shell/platform/android/io/flutter/embedding/engine/FlutterEnginePluginRegistry.java (hunk 12 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/engine/FlutterEnginePluginRegistry.java]: Talked offline, resolving old comments.

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java (hunk 11 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java]: Talked offline, resolving old comments.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java (hunk 10 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: Talked offline, resolving old comments.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java (hunk 9 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: Talked offline, resolving old comments.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java (hunk 8 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: Talked offline, resolving old comments.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java (hunk 7 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: Talked offline, resolving old comments.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java (hunk 6 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: Talked offline, resolving old comments.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java (hunk 5 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: Talked offline, resolving old comments.

File: shell/platform/android/io/flutter/embedding/android/FlutterFragment.java (hunk 4 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterFragment.java]: Talked offline, resolving old comments.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java (hunk 3 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: Talked offline, resolving old comments.

File: shell/platform/android/io/flutter/embedding/android/FlutterActivity.java (hunk 2 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterActivity.java]: Talked offline, resolving old comments.

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java (hunk 13 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java]: Yeah, with the way `PlatformViewsController` is integrated with the embedding right now, it's a requirement. If we'd like to make it optional then we'd need to make that change within the platform views subsystem, itself, because right now I'm just matching how platform views integrates with the old embedding.

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java (hunk 14 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java]: I'll attempt to make this change and see how it goes.

Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java]: Actually, @xster, now I remember why the `PlatformViewsController` isn't created inside `FlutterView` - platform view functionality needs to outlive a given `FlutterView`, right? When the engine is detached from a view, we probably won't attempt to layout or render anything, but we probably can't just tear out the platform views subsystem just because we stopped rendering to the screen, can we?

Comment (xster) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java]: Doesn't this apply to the FlutterFragment (to which you're giving ownership of the platform view controller in this PR) as well?

Comment (xster) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java]: Also, I think moving flutterEngine.getActivityControlSurface().attachToActivity() from onAttach to onCreateView is onActivityCreated is a bit orthogonal. I just noticed it now but it wasn't really part of this PR. We should make it cleaner nevertheless either in this PR or in another one.

Comment (matthew-carroll) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java]: @xster I just pushed a change that moves `PlatformViewsController` into `FlutterEngine` as a component. I think that conceptually makes sense since we've said that some notion of platform views needs to function even in the background. Let me know what you think.

WRT to `onActivityCreated`, we should chat about it. I looked up the Fragment lifecycle and I'm not convinced that there is a symmetric call. I think the question comes down to this: if I attach an existing Fragment to an existing Activity, will we get a call to `onActivityCreated`? If not, then I don't think we can use it.

**Comment by matthew-carroll** [comment]
Rebased to resolve conflicts and force pushed. Also added commit that implements a method needed for keyboard support. Verified that accessibility works the same way between the old and new embedding in the webview example project.

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java [hunk 15]
```diff
@@ -511,7 +541,9 @@ private void resetWillNotDraw(boolean isAccessibilityEnabled, boolean isTouchExp
    * See {@link #detachFromFlutterEngine()} for information on how to detach from a
    * {@link FlutterEngine}.
    */
-  public void attachToFlutterEngine(@NonNull FlutterEngine flutterEngine) {
+  public void attachToFlutterEngine(
```
Comment (xster) [inline shell/platform/android/io/flutter/embedding/android/FlutterView.java]: This whole thing looks a lot cleaner now. Thanks!
